### DISCUSSION TITLE
Create factory bin for ESP32 C2 Help

### DISCUSSION THREAD

**Discussion by Basti149** [arendst/Tasmota#20791] [Q&A]
First of all, a big compliment to the developers of Tasmota and especially that the ESPC2 is now supported. It works great with Tasmota despite the fact that it is beta. But the view in version 13.4.0.1 looks really good. 
Now I still have a question, I have flashed the ESPC2 with the Tasmota version "tasmota32c2-arduino30.factory.bin" which also worked.
Now I want to add two actuators via platformio, one is the PCA9685 and the other is the MCP23017, but unfortunately the C2 won't start after that.

It can't be due to the compilation as it also worked with an ESP8266. I have selected "tasmota32c2-arduino30" for the C2. Is it because it is not a factory.bin? 
And how do I create the factory bin with the MCP23017 and PCA9685?

//MCP23017 Unterstützung

#define USE_MCP230xx_ADDR 0x20
#define USE_MCP230xx // Enable INPUT mode (pinmode 1 through 4)
#define USE_MCP230xx_OUTPUT // Enable OUTPUT mode (pinmode 5)
#define USE_MCP230xx_DISPLAYOUTPUT // Display state of OUTPUT pins on main Tasmota web interface
#define USE_MCP23XXX_DRV //multiple switch/button/relay Modus
#endif

//PCA9685 Unterstützung

#ifndef USE_PCA9685
#define USE_PCA9685                     // [I2cDriver1] Enable PCA9685 I2C HW PWM Driver - Must define I2C Address in #define USE_PCA9685_ADDR below - range 0x40 - 0x47 (+1k4 code)
  #define USE_PCA9685_ADDR 0x40         // Enable PCA9685 I2C Address to use (Must be within range 0x40 through 0x47 - set according to your wired setup)
  #define USE_PCA9685_FREQ 50           // Define default PWM frequency in Hz to be used (must be within 24 to 1526) - If other value is used, it will revert to 50Hz
#endif

**Comment by sfromis** [comment]
What file did you try to upgrade to, and how? For a normal upgrade, you should not use a `.factory` binary, which is only for wired initial flashing. If that's what you want, you should find it under the directory `build_output`, along with the OTA binary.

**Comment by Basti149** [comment]
I tried it once via ota update (adapted version) directly on tasmota gui. But there was the message “Fail Upload” and saveboot in red. No idea if that's correct? If Platformio doesn't have a correct version, so far I've only compiled it myself with the ESP8266, which always works when I first flash it. Where do I get the c2 customized factory.bin?

Holen Sie sich Outlook für Android<https://aka.ms/ghei36>
________________________________

**Reply by sfromis** [reply]
Again, what file did you try to upgrade to?

For OTA upgrade you **cannot** use a `.factory` binary, but as I mentioned, you can find it under the folder `build_output` if you need it for wired flashing  "from cold".

**Comment by Basti149** [comment]
I flashed the C2 for the first time with the factory bin (it worked and c2 was accessible). And wanted to update the adapted version (of Platformio) with the two actuators via ota update. But that failed and it said "saveboot" in red text. what did i do wrong?




A second attempt was to create a bin for the adapted version for the ESP8266 and flash it, which worked without errors.



Mit freundlichen Grüßen Sebastian Köhler.
________________________________

**Comment by Jason2866** [comment]
esp32x and esp8266 are very different how to be flashed and OTA updated.
Nothing is common here. The esp32x needs beside the compiled firmware image other "helper" images.
To make this easy we provide a "factory" image where all the needed files are merged together in one file.
This file is flashed initially. If this is done correctly, Tasmota is running and can be updated OTA.
For this process only and just only the compiled Tasmota firmware Image is needed.
The firmware can easily flashed by uploading the firmware via the file upload button.
Tasmota switches to saveboot firmware. This firmware flashes the the provided new firmware.
If something went wrong safeboot firmware comes up again and the firmware can be provided there too.
Both firmware variants are provided in folder `build_output`

**Reply by sfromis** [reply]
The short version is that for all ESP32 variants (including the C2), you must use the `tasmota32c2.factory.bin` (etc) *only* for wired flashing, and the `tasmota32c2.bin` (etc) only for OTA upgrade via the web UI.

**Comment by Basti149** [comment]
Thank you so much I won't try that again. Which version should I compile as?

![grafik](https://github.com/arendst/Tasmota/assets/58861066/ed2dab4a-fbcb-48ec-b7c0-682da89f389d)


I think I need the "tasmota32c2-arduino30" right?

**Comment by Basti149** [comment]
Many thanks for your help. I managed it after deleting the whole project and re-downloading Tasmota. After a new compile, the ota update also worked and the C2 found it.

Thank you both very much and also for the great integration of the C2 and also really good with the new feature “Gpio Viewer”.

The only thing you can do is increase the frequencies a little with the PCA9685 because it doesn't dim as smoothly as with the ESP's GPIOS, but these are just small things.

**Comment by Jason2866** [comment]
> The only thing you can do is increase the frequencies a little with the PCA9685 because it doesn't dim as smoothly as with the ESP's GPIOS, but these are just small things.

There should be no difference whatever esp32 MCU is used. PWM frequency is in hardware implemented.
The MCU just get "told" which frequency is wanted.

**Reply by Basti149** [reply]
no MCU, I meant the PCA9685

**Reply by sfromis** [reply]
Whatever term, Tasmota can only tell the PCA9685 what frequency is wanted. When you do a build to enable the PCA9685 driver, you can also config the default PWM frequency:
https://github.com/arendst/Tasmota/blob/development/tasmota/my_user_config.h#L630-L633

You can also use the command `Driver15` to set frequency. Comments for the V2 variant of the driver indicates what arguments you can use:
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_15_pca9685_v2.ino#L28-L75

**Comment by Basti149** [comment]
Thanks, I'll try it tomorrow. The frequency is currently 50, according to the specifications the PCA can handle even more.


First of all, I'm happy that the C2 works so easily with Tasmota. Despite the beta version

Mit freundlichen Grüßen Sebastian Köhler.
________________________________
